File: MainProject/Code.Domains.Models/ValueObjects/PlanetGenerates/HexTileDataOverrider.cs
using Commons.Enums;
using Commons.Utils;
using Domains.Models.Entities.PlanetGenerates;

namespace Domains.Models.ValueObjects.PlanetGenerates;

public enum OptionalToggle
{
    Ignore,
    Yes,
    No
}

public struct HexTileDataOverrider
{
    public HexTileDataOverrider()
    {
    }

    public bool EditMode;
    public bool ApplyTerrain;
    public int ActiveTerrain;
    public bool ApplyElevation;
    public int ActiveElevation;
    public bool ApplyWaterLevel;
    public int ActiveWaterLevel;
    public int BrushSize;
    public OptionalToggle RiverMode;
    public OptionalToggle RoadMode;
    public bool ApplyUrbanLevel;
    public int ActiveUrbanLevel;
    public bool ApplyFarmLevel;
    public int ActiveFarmLevel;
    public bool ApplyPlantLevel;
    public int ActivePlantLevel;
    public OptionalToggle WalledMode;
    public bool ApplySpecialIndex;
    public int ActiveSpecialIndex;
    public HashSet<Tile> OverrideTiles = [];

    public bool IsOverridingTileConnection(Tile tile, Tile neighbor) =>
        EditMode && OverrideTiles.Count > 0
                 && OverrideTiles.Contains(tile) && !OverrideTiles.Contains(neighbor);

    public bool IsOverrideTile(Tile tile) => EditMode && OverrideTiles.Contains(tile);
    public bool IsOverrideNoRiver(Tile tile) => IsOverrideTile(tile) && RiverMode == OptionalToggle.No;
    public bool IsOverrideNoRoad(Tile tile) => IsOverrideTile(tile) && RoadMode == OptionalToggle.No;

    public int Elevation(Tile tile) =>
        IsOverrideTile(tile) && ApplyElevation ? ActiveElevation : tile.Data.Elevation;

    public int WaterLevel(Tile tile) =>
        IsOverrideTile(tile) && ApplyWaterLevel ? ActiveWaterLevel : tile.Data.WaterLevel;

    public bool IsUnderwater(Tile tile) =>
        IsOverrideTile(tile) ? WaterLevel(tile) > Elevation(tile) : tile.Data.IsUnderwater;

    public float StreamBedY(Tile tile, float unitHeight) =>
        IsOverrideTile(tile)
            ? (Elevation(tile) + HexMetrics.StreamBedElevationOffset) * unitHeight
            : tile.Data.StreamBedY(unitHeight);

    public float RiverSurfaceY(Tile tile, float unitHeight) =>
        IsOverrideTile(tile)
            ? (Elevation(tile) + HexMetrics.WaterElevationOffset) * unitHeight
            : tile.Data.RiverSurfaceY(unitHeight);

    public float WaterSurfaceY(Tile tile, float unitHeight) =>
        IsOverrideTile(tile)
            ? (WaterLevel(tile) + HexMetrics.WaterElevationOffset) * unitHeight
            : tile.Data.WaterSurfaceY(unitHeight);

    public bool HasRiver(Tile tile) => !IsOverrideNoRiver(tile) && tile.Data.HasRiver;
    public bool HasIncomingRiver(Tile tile) => !IsOverrideNoRiver(tile) && tile.Data.HasIncomingRiver;
    public bool HasOutgoingRiver(Tile tile) => !IsOverrideNoRiver(tile) && tile.Data.HasOutgoingRiver;
    public bool HasRiverBeginOrEnd(Tile tile) => !IsOverrideNoRiver(tile) && tile.Data.HasRiverBeginOrEnd;

    public bool HasRiverThroughEdge(Tile tile, int idx) =>
        !IsOverrideNoRiver(tile) && tile.Data.HasRiverThroughEdge(idx);

    public bool HasIncomingRiverThroughEdge(Tile tile, int idx) =>
        !IsOverrideNoRiver(tile) && tile.Data.HasIncomingRiverThroughEdge(idx);

    public bool HasRoads(Tile tile) => !IsOverrideNoRoad(tile) && tile.Data.HasRoads;

    public bool HasRoadThroughEdge(Tile tile, int idx) =>
        !IsOverrideNoRoad(tile) && tile.Data.HasRoadThroughEdge(idx);

    public bool Walled(Tile tile) =>
        IsOverrideTile(tile) && WalledMode != OptionalToggle.Ignore
            ? WalledMode == OptionalToggle.Yes
            : tile.Data.Walled;

    public HexEdgeType GetEdgeType(Tile tile1, Tile tile2) =>
        HexMetrics.GetEdgeType(Elevation(tile1), Elevation(tile2));

    public int UrbanLevel(Tile tile) =>
        IsOverrideTile(tile) && ApplyUrbanLevel ? ActiveUrbanLevel : tile.Data.UrbanLevel;

    public int FarmLevel(Tile tile) =>
        IsOverrideTile(tile) && ApplyFarmLevel ? ActiveFarmLevel : tile.Data.FarmLevel;

    public int PlantLevel(Tile tile) =>
        IsOverrideTile(tile) && ApplyPlantLevel ? ActivePlantLevel : tile.Data.PlantLevel;

    public int SpecialIndex(Tile tile) =>
        IsOverrideTile(tile) && ApplySpecialIndex ? ActiveSpecialIndex : tile.Data.SpecialIndex;

    public bool IsSpecial(Tile tile) => SpecialIndex(tile) > 0;
}

File: TerraObserver/TO.FSharp/Repos/Models/Meshes/HexTileDataOverrider.fs
namespace TO.FSharp.Repos.Models.Meshes

open System.Collections.Generic
open TO.FSharp.Commons.Utils
open TO.FSharp.Repos.Models.HexSpheres.Tiles

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-09 17:19:09
type OptionalToggle =
    | Ignore = 0
    | Yes = 1
    | No = 2

type HexTileDataOverrider() =
    [<DefaultValue>] val mutable EditMode: bool
    [<DefaultValue>] val mutable ApplyTerrain: bool
    [<DefaultValue>] val mutable ActiveTerrain: int
    [<DefaultValue>] val mutable ApplyElevation: bool
    [<DefaultValue>] val mutable ActiveElevation: int
    [<DefaultValue>] val mutable ApplyWaterLevel: bool
    [<DefaultValue>] val mutable ActiveWaterLevel: int
    [<DefaultValue>] val mutable BrushSize: int
    [<DefaultValue>] val mutable RiverMode: OptionalToggle
    [<DefaultValue>] val mutable RoadMode: OptionalToggle
    [<DefaultValue>] val mutable ApplyUrbanLevel: bool
    [<DefaultValue>] val mutable ActiveUrbanLevel: int
    [<DefaultValue>] val mutable ApplyFarmLevel: bool
    [<DefaultValue>] val mutable ActiveFarmLevel: int
    [<DefaultValue>] val mutable ApplyPlantLevel: bool
    [<DefaultValue>] val mutable ActivePlantLevel: int
    [<DefaultValue>] val mutable WalledMode: OptionalToggle
    [<DefaultValue>] val mutable ApplySpecialIndex: bool
    [<DefaultValue>] val mutable ActiveSpecialIndex: int
    member val OverrideTileIds = HashSet<int>()

    member this.IsOverridingTileConnection (tile: int) (neighbor: int) =
        this.EditMode
        && this.OverrideTileIds.Count > 0
        && this.OverrideTileIds.Contains tile
        && not <| this.OverrideTileIds.Contains neighbor

    member this.IsOverrideTile(tile: int) =
        this.EditMode && this.OverrideTileIds.Contains tile

    member this.IsOverrideNoRiver(tile: int) =
        this.IsOverrideTile tile && this.RiverMode = OptionalToggle.No

    member this.IsOverrideNoRoad(tile: int) =
        this.IsOverrideTile tile && this.RoadMode = OptionalToggle.No

    member this.Elevation (tile: int) (tileValue: TileValue) =
        if this.IsOverrideTile tile && this.ApplyElevation then
            this.ActiveElevation
        else
            tileValue.Elevation

    member this.GetEdgeType (tile1: int) (tileValue1: TileValue) (tile2: int) (tileValue2: TileValue) =
        HexMetrics.getEdgeType
        <| this.Elevation tile1 tileValue1
        <| this.Elevation tile2 tileValue2

    member this.WaterLevel (tile: int) (tileValue: TileValue) =
        if this.IsOverrideTile tile && this.ApplyWaterLevel then
            this.ActiveWaterLevel
        else
            tileValue.WaterLevel

    member this.IsUnderwater (tile: int) (tileValue: TileValue) =
        if this.IsOverrideTile tile then
            this.WaterLevel tile tileValue > this.Elevation tile tileValue
        else
            tileValue.IsUnderwater

    member this.StreamBedY (tile: int) (tileValue: TileValue) (unitHeight: float32) =
        if this.IsOverrideTile tile then
            (float32 (this.Elevation tile tileValue) + HexMetrics.streamBedElevationOffset)
            * unitHeight
        else
            tileValue.StreamBedY unitHeight

    member this.RiverSurfaceY (tile: int) (tileValue: TileValue) (unitHeight: float32) =
        if this.IsOverrideTile tile then
            (float32 (this.Elevation tile tileValue) + HexMetrics.waterElevationOffset)
            * unitHeight
        else
            tileValue.StreamBedY unitHeight

    member this.WaterSurfaceY (tile: int) (tileValue: TileValue) (unitHeight: float32) =
        if this.IsOverrideTile tile then
            (float32 (this.WaterLevel tile tileValue) + HexMetrics.waterElevationOffset)
            * unitHeight
        else
            tileValue.WaterSurfaceY unitHeight

    member this.HasRivers (tile: int) (tileFlag: TileFlag) =
        not <| this.IsOverrideNoRiver tile && tileFlag.HasRivers

    member this.HasIncomingRiver (tile: int) (tileFlag: TileFlag) =
        not <| this.IsOverrideNoRiver tile && tileFlag.HasIncomingRiver

    member this.HasOutgoingRiver (tile: int) (tileFlag: TileFlag) =
        not <| this.IsOverrideNoRiver tile && tileFlag.HasOutgoingRiver

    member this.HasRiverBeginOrEnd (tile: int) (tileFlag: TileFlag) =
        not <| this.IsOverrideNoRiver tile && tileFlag.HasRiverBeginOrEnd

    member this.HasRiverThroughEdge (tile: int) (tileFlag: TileFlag) (dir: int) =
        not <| this.IsOverrideNoRiver tile && tileFlag.HasRiverThroughEdge dir

    member this.HasIncomingRiverThroughEdge (tile: int) (tileFlag: TileFlag) (dir: int) =
        not <| this.IsOverrideNoRiver tile && tileFlag.HasIncomingRiverThoughEdge dir

    member this.HasRoads (tile: int) (tileFlag: TileFlag) =
        not <| this.IsOverrideNoRoad tile && tileFlag.HasRoads

    member this.HasRoadThroughEdge (tile: int) (tileFlag: TileFlag) (dir: int) =
        not <| this.IsOverrideNoRoad tile && tileFlag.HasRoadThroughEdge dir

    member this.Walled (tile: int) (tileFlag: TileFlag) =
        if this.IsOverrideTile tile && this.WalledMode <> OptionalToggle.Ignore then
            this.WalledMode = OptionalToggle.Yes
        else
            tileFlag.Walled

    member this.UrbanLevel (tile: int) (tileValue: TileValue) =
        if this.IsOverrideTile tile && this.ApplyUrbanLevel then
            this.ActiveUrbanLevel
        else
            tileValue.UrbanLevel

    member this.FarmLevel (tile: int) (tileValue: TileValue) =
        if this.IsOverrideTile tile && this.ApplyFarmLevel then
            this.ActiveFarmLevel
        else
            tileValue.FarmLevel

    member this.PlantLevel (tile: int) (tileValue: TileValue) =
        if this.IsOverrideTile tile && this.ApplyPlantLevel then
            this.ActivePlantLevel
        else
            tileValue.PlantLevel

    member this.SpecialIndex (tile: int) (tileValue: TileValue) =
        if this.IsOverrideTile tile && this.ApplySpecialIndex then
            this.ActiveSpecialIndex
        else
            tileValue.SpecialIndex

    member this.IsSpecial (tile: int) (tileValue: TileValue) = this.SpecialIndex tile tileValue <> 0
